File: AlgoTest/src/algorithm/Arena.java
package algorithm;

class Arena {
    static final int GRID_UNKNOWN = 0;
    static final int GRID_EXPLORED = 1;
    static final int GRID_SUSPECT = 2;
    static final int GRID_OBSTACLE = 3;

    static final int FORWARD = 0;
    static final int LEFT = 1;
    static final int RIGHT = 2;
    private static final int UP = 0;
    private static final int DOWN = 3;

    private int[][] gridArray = new int[20][15];
    private int[][] exploreArray = new int[20][15];
    private int[][] obstacleArray = new int[20][15];
    private int facing = UP;

    private Coordinates currentPosition;
    private Coordinates startPosition;
    private Coordinates goalPosition;
    private Coordinates waypointPosition;

    Arena() {
        startPosition = new Coordinates(1, 1);
        goalPosition = new Coordinates(13, 18);
        currentPosition = new Coordinates(1, 1);
        waypointPosition = new Coordinates(-1, -1);
        updateCurrentPosition(startPosition.x, startPosition.y);
    }

    void move(int direction, int distance) {
        switch (direction) {
            case FORWARD:
                switch (facing) {
                    case UP:
                        updateCurrentPosition(currentPosition.x, currentPosition.y + distance);
                        break;
                    case LEFT:
                        updateCurrentPosition(currentPosition.x - distance, currentPosition.y);
                        break;
                    case RIGHT:
                        updateCurrentPosition(currentPosition.x + distance, currentPosition.y);
                        break;
                    case DOWN:
                        updateCurrentPosition(currentPosition.x, currentPosition.y - distance);
                        break;
                } break;

            case LEFT:
                switch (facing) {
                    case UP:
                        facing = LEFT;
                        break;
                    case LEFT:
                        facing = DOWN;
                        break;
                    case RIGHT:
                        facing = UP;
                        break;
                    case DOWN:
                        facing = RIGHT;
                        break;
                } break;

            case RIGHT:
                switch (facing) {
                    case UP:
                        facing = RIGHT;
                        break;
                    case LEFT:
                        facing = UP;
                        break;
                    case RIGHT:
                        facing = DOWN;
                        break;
                    case DOWN:
                        facing = LEFT;
                        break;
                } break;
        }
    }

    boolean isMovable(int direction) {
        int x = currentPosition.x;
        int y = currentPosition.y;

        switch (direction) {
            case FORWARD:
                switch (facing) {
                    case UP:
                        y += 1;
                        break;
                    case LEFT:
                        x -= 1;
                        break;
                    case RIGHT:
                        x += 1;
                        break;
                    case DOWN:
                        y -= 1;
                        break;
                } break;

            case LEFT:
                switch (facing) {
                    case UP:
                        x -= 1;
                        break;
                    case LEFT:
                        y -= 1;
                        break;
                    case RIGHT:
                        y += 1;
                        break;
                    case DOWN:
                        x += 1;
                        break;
                } break;

            case RIGHT:
                switch (facing) {
                    case UP:
                        x += 1;
                        break;
                    case LEFT:
                        y += 1;
                        break;
                    case RIGHT:
                        y -= 1;
                        break;
                    case DOWN:
                        x -= 1;
                        break;
                } break;
        }

        for (int yOffset = -1; yOffset <= 1; yOffset++) {
            for (int xOffset = -1; xOffset <= 1; xOffset++) {
                if (isInvalidCoordinates(x + xOffset, y + yOffset, false)) return false;
                if (obstacleArray[y + yOffset][x + xOffset] == 1) return false;
            }
        }

        return true;
    }

    void setGridSensor(int sensor, int distance, int status) {
        switch (sensor) {
            case 1:
            case 2:
            case 3:
                setGridSensorFront(sensor, distance, status);
                break;
            case 4:
            case 5:
                System.out.println("SENSOR"+sensor);
                setGridSensorLeft(sensor, distance, status);
                break;
            case 6:
                setGridSensorRight(distance, status);
                break;
        }
    }

    private void setGridSensorFront(int sensor, int distance, int status) {
        int x = currentPosition.x;
        int y = currentPosition.y;

        switch (facing) {
            case UP:
                x += (sensor - 2);
                y += 1;
                setGridStatusInternal(x, y + distance, status);
                for (int offset = distance - 1; offset > 0; offset--) setGridStatusInternal(x, y + offset, GRID_EXPLORED);
                break;
            case LEFT:
                x -= 1;
                y += (sensor - 2);
                setGridStatusInternal(x - distance, y, status);
                for (int offset = distance - 1; offset > 0; offset--) setGridStatusInternal(x - offset, y, GRID_EXPLORED);
                break;
            case RIGHT:
                x += 1;
                y -= (sensor - 2);
                setGridStatusInternal(x + distance, y, status);
                for (int offset = distance - 1; offset > 0; offset--) setGridStatusInternal(x + offset, y, GRID_EXPLORED);
                break;
            case DOWN:
                x -= (sensor - 2);
                y -= 1;
                setGridStatusInternal(x, y - distance, status);
                for (int offset = distance - 1; offset > 0; offset--) setGridStatusInternal(x, y - offset, GRID_EXPLORED);
                break;
        }
    }

    private void setGridSensorLeft(int sensor, int distance, int status) {
        int x = currentPosition.x;
        int y = currentPosition.y;

        switch (facing) {
            case UP:
                x -= 1;
                y -= (sensor - 4);
                setGridStatusInternal(x - distance, y, status);
                for (int offset = distance - 1; offset > 0; offset--) setGridStatusInternal(x - offset, y, GRID_EXPLORED);
                break;
            case LEFT:
                x += (sensor - 4);
                y -= 1;
                setGridStatusInternal(x, y - distance, status);
                for (int offset = distance - 1; offset > 0; offset--) setGridStatusInternal(x, y - offset, GRID_EXPLORED);
                break;
            case RIGHT:
                x -= (sensor - 4);
                y += 1;
                setGridStatusInternal(x, y + distance, status);
                for (int offset = distance - 1; offset > 0; offset--) setGridStatusInternal(x, y + offset, GRID_EXPLORED);
                break;
            case DOWN:
                x += 1;
                y += (sensor - 4);
                setGridStatusInternal(x + distance, y, status);
                for (int offset = distance - 1; offset > 0; offset--) setGridStatusInternal(x + offset, y, GRID_EXPLORED);
                break;
        }
    }

    private void setGridSensorRight(int distance, int status) {
        int x = currentPosition.x;
        int y = currentPosition.y;

        switch (facing) {
            case UP:
                x += 1;
                setGridStatusInternal(x + distance, y, status);
                for (int offset = distance - 1; offset > 0; offset--) setGridStatusInternal(x + offset, y, GRID_EXPLORED);
                break;
            case LEFT:
                y += 1;
                setGridStatusInternal(x, y + distance, status);
                for (int offset = distance - 1; offset > 0; offset--) setGridStatusInternal(x, y + offset, GRID_EXPLORED);
                break;
            case RIGHT:
                y -= 1;
                setGridStatusInternal(x, y - distance, status);
                for (int offset = distance - 1; offset > 0; offset--) setGridStatusInternal(x, y - offset, GRID_EXPLORED);
                break;
            case DOWN:
                x -= 1;
                setGridStatusInternal(x - distance, y, status);
                for (int offset = distance - 1; offset > 0; offset--) setGridStatusInternal(x - offset, y, GRID_EXPLORED);
                break;
        }
    }

    private void updateCurrentPosition(int x, int y) {
        if (isInvalidCoordinates(x, y, true)) return;
        currentPosition.x = x;
        currentPosition.y = y;

        for (int yOffset = -1; yOffset <= 1; yOffset++) {
            for (int xOffset = -1; xOffset <= 1; xOffset++) {
                setGridStatusInternal(x + xOffset, y + yOffset, GRID_EXPLORED);
            }
        }
    }

    void displayGrid() {
        System.out.println();
        for (int y = 19; y >= 0; y--) {
            for (int x = 0; x <= 14; x++) {
                int status = gridArray[y][x];

                switch (status) {
                    case GRID_EXPLORED:
                        System.out.print("O ");
                        break;

                    case GRID_SUSPECT:
                        System.out.print("x ");
                        break;

                    case GRID_OBSTACLE:
                        System.out.print("X ");
                        break;

                    default:
                        System.out.print("? ");
                        break;
                }
            }

            System.out.println();
        }
        System.out.println();
    }

    private void setGridStatusInternal(int x, int y, int status) {
        if (isInvalidCoordinates(x, y, false)) return;
        if (gridArray[y][x] == GRID_OBSTACLE) return;
        gridArray[y][x] = status;

        if (status != GRID_UNKNOWN) exploreArray[y][x] = 1;
        if (status == GRID_OBSTACLE) obstacleArray[y][x] = 1;
    }

    private boolean isInvalidCoordinates(int x, int y, boolean robot) {
        if (robot) return (x < 1 || x > 13 || y < 1 || y > 18);
        else return (x < 0 || x > 14 || y < 0 || y > 19);
    }

    private class Coordinates {
        private int x;
        private int y;

        private Coordinates(int x, int y) {
            this.x = x;
            this.y = y;
        }
    }
}

File: Algorithms/src/main/kotlin/wjayteo/mdp/algorithms/arena/Arena.kt
package wjayteo.mdp.algorithms.arena

import wjayteo.mdp.algorithms.algorithm.Algorithm.Companion.COVERAGE_LIMIT
import wjayteo.mdp.algorithms.uicomponent.ArenaMapView
import wjayteo.mdp.algorithms.uicomponent.MasterView
import wjayteo.mdp.algorithms.wifi.WifiSocketController
import kotlin.math.exp

class Arena {
    companion object {
        private const val GRID_UNKNOWN      : Int = 0
        private const val GRID_EXPLORED     : Int = 1
        private const val GRID_SUSPECT      : Int = 2
        private const val GRID_OBSTACLE     : Int = 3

        val start = Coordinates(1, 1)
        val goal = Coordinates(13, 18)
        val waypoint = Coordinates(-1, -1)
        val exploreArray: Array<Array<Int>> = Array(20) { Array(15) { 0 } }
        val obstacleArray: Array<Array<Int>> = Array(20) { Array(15) { 0 } }
        private val gridStateArray: Array<Array<Int>> = Array(20) { Array(15) { 0 } }
        private val visitedArray: Array<Array<Int>> = Array(20) { Array(15) { 0 } }
        private val scannedArray: Array<Array<Int>> = Array(20) { Array(15) { 0 } }
        private val counterArray: Array<Array<Int>> = Array(20) { Array(15) { 0 } }
        private var attachedView: ArenaMapView? = null

        fun setAttachedView(view: ArenaMapView) {
            attachedView = view
        }

        fun reset() {
            for (y in 19 downTo 0) for (x in 0..14) setUnknown(x, y)
            waypoint.x = -1
            waypoint.y = -1
            attachedView?.removeWaypoint()
        }

        fun setScanned(x: Int, y: Int) {
            if (isInvalidCoordinates(x, y)) return

            if (isObstacle(x, y)) {
                scannedArray[y][x] = 1
                attachedView?.setScanned(x, y)
            }
        }

        fun isScanned(x: Int, y: Int): Boolean {
            if (isInvalidCoordinates(x, y)) return true
            if (!isObstacle(x, y)) return true
            return (scannedArray[y][x] == 1)
        }


        fun setVisited(x: Int, y: Int) {
            if (isInvalidCoordinates(x, y)) return
            visitedArray[y][x]++
        }

        fun isVisitedRepeated(x: Int, y: Int): Boolean {
            return (visitedArray[y][x] >= 3)
        }

        fun loadMap(explore: Array<Array<Int>>, obstacle: Array<Array<Int>>) {
            reset()

            for (y in 19 downTo 0) {
                for (x in 0..14) {
                    val exploreBit: Int = explore[y][x]
                    val obstacleBit: Int = obstacle[y][x]
                    exploreArray[y][x] = exploreBit
                    obstacleArray[y][x] = obstacleBit
                    if (exploreBit == 1) setExplored(x, y)
                    if (obstacleBit == 1) setObstacle(x, y)
                }
            }
        }

        fun isInvalidCoordinates(x: Int, y: Int, robotSize: Boolean = false): Boolean {
            return if (robotSize) (x < 1 || x > 13 || y < 1 || y > 18)
            else (x < 0 || x > 14 || y < 0 || y > 19)
        }

        fun isInvalidCoordinates(coordinates: Coordinates, robotSize: Boolean = false): Boolean {
            val x: Int = coordinates.x
            val y: Int = coordinates.y
            return isInvalidCoordinates(x, y, robotSize)
        }

        fun isExplored(x: Int, y: Int): Boolean {
            if (isInvalidCoordinates(x, y)) return true
            return (exploreArray[y][x] == 1)
        }

        fun setAllExplored() {
            for (y in 19 downTo 0) {
                for (x in 0..14) {
                    setExplored(x, y)
                }
            }
        }

        fun setAllUnknown() {
            for (y in 19 downTo 0) {
                for (x in 0..14) {
                    if (obstacleArray[y][x] == 0) setUnknown(x, y)
                }
            }

            Robot.move(Robot.position.x, Robot.position.y)
        }

        fun isEveryGridExplored(): Boolean {
            for (y in 19 downTo 0) {
                for (x in 0..14) {
                    if (exploreArray[y][x] == 0) return false
                }
            }

            return true
        }

        fun isMovable(x: Int, y: Int): Boolean {
            if (isInvalidCoordinates(x, y, true)) return false

            for (yOffset in -1 .. 1) {
                for (xOffset in -1 .. 1) {
                    if (isObstacle(x + xOffset, y + yOffset)) return false
                }
            }

            return true
        }

        fun setStartPoint(x: Int, y: Int): Boolean {
            if (isInvalidCoordinates(x, y)) return false
            if (x == start.x && y == start.y) return true

            for (yOffset in -1..1) {
                for (xOffset in -1..1) {
                    val x1: Int = x + xOffset
                    val y1: Int = y + yOffset
                    if (isObstacle(x1, y1) || isGoalPoint(x1, y1) || isWaypoint(x1, y1)) return false
                }
            }

            start.x = x
            start.y = y
            setExploredPoint(x, y)
            attachedView?.setStartPoint(x, y)
            return true
        }

        fun setGoalPoint(x: Int, y: Int): Boolean {
            if (isInvalidCoordinates(x, y)) return false
            if (x == goal.x && y == goal.y) return true

            for (yOffset in -1..1) {
                for (xOffset in -1..1) {
                    val x1: Int = x + xOffset
                    val y1: Int = y + yOffset
                    if (isObstacle(x1, y1) || isStartPoint(x1, y1) || isWaypoint(x1, y1)) return false
                }
            }

            goal.x = x
            goal.y = y
            setExploredPoint(x, y)
            attachedView?.setGoalPoint(x, y)
            return true
        }

        fun setWaypoint(x: Int, y: Int): Boolean {
            if (isInvalidCoordinates(x, y)) return false
            if (x == waypoint.x && y == waypoint.y) return true

            for (yOffset in -1..1) {
                for (xOffset in -1..1) {
                    val x1: Int = x + xOffset
                    val y1: Int = y + yOffset
                    if (isObstacle(x1, y1) || isGoalPoint(x1, y1) || isStartPoint(x1, y1)) return false
                }
            }

            waypoint.x = x
            waypoint.y = y
            setExploredPoint(x, y)
            attachedView?.setWaypoint(x, y)
            return true
        }

        private fun setExploredPoint(x: Int, y: Int) {
            for (yOffset in -1..1) {
                for (xOffset in -1..1) {
                    val x1: Int = x + xOffset
                    val y1: Int = y + yOffset
                    setExploredForced(x1, y1)
                }
            }
        }

        fun refreshPoints() {
            setExploredPoint(waypoint.x, waypoint.y)
            setExploredPoint(start.x, start.y)
            setExploredPoint(goal.x, goal.y)
        }

        fun setExplored(x: Int, y: Int) {
            if (isInvalidCoordinates(x, y)) return
            if (gridStateArray[y][x] >= GRID_OBSTACLE) return
            gridStateArray[y][x] = GRID_EXPLORED
            exploreArray[y][x] = 1
            attachedView?.setExplored(x, y)
        }

        fun setExploredForced(x: Int, y: Int) {
            if (isInvalidCoordinates(x, y)) return
            gridStateArray[y][x] = GRID_EXPLORED
            exploreArray[y][x] = 1
            obstacleArray[y][x] = 0
            attachedView?.setExplored(x, y)
        }

        fun endHug() {
            for (y in 19 downTo 0) {
                for (x in 0..14) {
                    if (gridStateArray[y][x] == GRID_SUSPECT) setObstacle(x, y)
                    if (gridStateArray[y][x] == GRID_UNKNOWN) setExplored(x, y)
                    if (visitedArray[y][x] > 0) setExploredForced(x, y)
                }
            }
        }

        fun setGridState(x: Int, y: Int, value: Int) {
            if (isInvalidCoordinates(x, y) || isGoalPoint(x, y) || isStartPoint(x, y) || isWaypoint(x, y)) return
            counterArray[y][x] += value

            if (counterArray[y][x] > 0) {
                gridStateArray[y][x] = GRID_OBSTACLE
                exploreArray[y][x] = 1
                obstacleArray[y][x] = 1
                if (scannedArray[y][x] == 1) attachedView?.setScanned(x, y)
                else attachedView?.setObstacle(x, y)
            } else {
                gridStateArray[y][x] = GRID_EXPLORED
                exploreArray[y][x] = 1
                obstacleArray[y][x] = 0
                scannedArray[y][x] = 0
                attachedView?.setExplored(x, y)
            }
        }

        private fun setObstacle(x: Int, y: Int) {
            if (isInvalidCoordinates(x, y)) return
            if (gridStateArray[y][x] >= GRID_OBSTACLE) return

            if (isOccupied(x, y)) {
                setExplored(x, y)
                return
            }

            gridStateArray[y][x] = GRID_OBSTACLE
            exploreArray[y][x] = 1
            obstacleArray[y][x] = 1
            attachedView?.setObstacle(x, y)
        }

        fun plotFastestPath() {
            for (y in 19 downTo 0) {
                for (x in 0..14) {
                    if (isObstacle(x, y)) continue
                    if (isExplored(x, y)) setExplored(x, y) else setUnknown(x, y)
                }
            }

            val pathList: List<IntArray> = MasterView.fastestPath.computeFastestPath()

            for (p in pathList) {
                val x = p[0]
                val y = p[1]
                setFastestPath(x, y)
            }
        }

        fun sendArena() {
            val descriptor: List<String> = MapDescriptor.fromArray(exploreArray, obstacleArray, 1)
            WifiSocketController.write("D", "#r:${descriptor[0]},${descriptor[1]},${Robot.position.x},${Robot.position.y},${Robot.facing}")
        }

        fun plotObstacle(x: Int, y: Int) {
            if (isInvalidCoordinates(x, y)) return

            if (isObstacle(x, y)) {
                setUnknown(x, y)
                setExplored(x, y)
                return
            }

            if (isOccupied(x, y)) return
            setObstacle(x, y)
        }

        fun coverageReached(): Boolean {
            var coveredCount = 0
            exploreArray.forEach { it.forEach { grid -> if (grid == 1) coveredCount++ } }
            return ((1.0 * coveredCount / 300) * 100 >= COVERAGE_LIMIT)
        }

        private fun setFastestPath(x: Int, y: Int) {
            if (isInvalidCoordinates(x, y)) return
            if (gridStateArray[y][x] >= GRID_OBSTACLE) return
            attachedView?.setFastestPath(x, y)
        }

        private fun setUnknown(x: Int, y: Int) {
            if (isInvalidCoordinates(x, y)) return
            gridStateArray[y][x] = GRID_UNKNOWN
            exploreArray[y][x] = 0
            obstacleArray[y][x] = 0
            scannedArray[y][x] = 0
            visitedArray[y][x] = 0
            attachedView?.setUnknown(x, y)
        }

        private fun isOccupied(x: Int, y: Int): Boolean {
            if (isInvalidCoordinates(x, y, false)) return true
            if (gridStateArray[y][x] >= GRID_OBSTACLE) return true
            if (isStartPoint(x, y) || isGoalPoint(x, y) || isWaypoint(x, y)) return true
            return false
        }

        fun isObstacle(x: Int, y: Int): Boolean {
            if (isInvalidCoordinates(x, y, false)) return true
            return (obstacleArray[y][x] == 1)
        }

        fun isObstacle2(x: Int, y: Int): Boolean {
            if (isInvalidCoordinates(x, y, false)) return false
            return (obstacleArray[y][x] == 1)
        }

        private fun isStartPoint(x: Int, y: Int): Boolean {
            if (isInvalidCoordinates(x, y, false)) return true
            if (isInvalidCoordinates(start)) return false
            return (intArrayOf(start.x - 1, start.x, start.x + 1).contains(x) && intArrayOf(start.y - 1, start.y, start.y + 1).contains(y))
        }

        private fun isGoalPoint(x: Int, y: Int): Boolean {
            if (isInvalidCoordinates(x, y, false)) return true
            if (isInvalidCoordinates(goal)) return false
            return (intArrayOf(goal.x - 1, goal.x, goal.x + 1).contains(x) && intArrayOf(goal.y - 1, goal.y, goal.y + 1).contains(y))
        }

        private fun isWaypoint(x: Int, y: Int): Boolean {
            if (isInvalidCoordinates(x, y, false)) return true
            if (isInvalidCoordinates(waypoint)) return false
            return (intArrayOf(waypoint.x - 1, waypoint.x, waypoint.x + 1).contains(x) && intArrayOf(waypoint.y - 1, waypoint.y, waypoint.y + 1).contains(y))
        }
    }
}